### PR TITLE
Fix for creating AGs on docker

### LINKED ISSUE #6814: Connect-DbaInstance to docker containers / Trying to get Get-DbaEndpoint to work with docker
I did some research today.
I have a lab with virtual maschines WIN10 (windows 10, i run dbatools there locally) and SRV1 (windows server 2016 core)
I have a laptop with docker and these two containers: https://dbatools.io/docker/

I run this on WIN10:

```
$DbaInstance = [DbaInstance[]]('.\SQL2017', 'localhost\SQL2017', 'WIN10\SQL2017', 'WIN10.Company.Pri\SQL2017', 'SRV1\SQL2017', 'SRV1.Company.Pri\SQL2017')
$DbaInstance | Format-Table
$doesNotWork, $DbaInstance = $DbaInstance  # First entry must be deleted
Connect-DbaInstance -SqlInstance $DbaInstance | Format-Table -Property Name, ComputerName, ComputerNamePhysicalNetBIOS, NetName, Platform

<# Output:

ComputerName      InstanceName Port NetworkProtocol IsLocalHost FullName                  FullSmoName               SqlComputerName     SqlInstanceName SqlFullName                
------------      ------------ ---- --------------- ----------- --------                  -----------               ---------------     --------------- -----------                
.\SQL2017         SQL2017         0              NP       False .\SQL2017\SQL2017         NP:.\SQL2017\SQL2017      [.\SQL2017]         [SQL2017]       [.\SQL2017\SQL2017]        
localhost         SQL2017         0             Any        True localhost\SQL2017         localhost\SQL2017         [localhost]         [SQL2017]       [localhost\SQL2017]        
WIN10             SQL2017         0             Any        True WIN10\SQL2017             WIN10\SQL2017             [WIN10]             [SQL2017]       [WIN10\SQL2017]            
WIN10.Company.Pri SQL2017         0             Any        True WIN10.Company.Pri\SQL2017 WIN10.Company.Pri\SQL2017 [WIN10.Company.Pri] [SQL2017]       [WIN10.Company.Pri\SQL2017]
SRV1              SQL2017         0             Any       False SRV1\SQL2017              SRV1\SQL2017              [SRV1]              [SQL2017]       [SRV1\SQL2017]             
SRV1.Company.Pri  SQL2017         0             Any       False SRV1.Company.Pri\SQL2017  SRV1.Company.Pri\SQL2017  [SRV1.Company.Pri]  [SQL2017]       [SRV1.Company.Pri\SQL2017] 

Name                      ComputerName ComputerNamePhysicalNetBIOS NetName Platform
----                      ------------ --------------------------- ------- --------
localhost\SQL2017         WIN10        WIN10                       WIN10   NT x64  
WIN10\SQL2017             WIN10        WIN10                       WIN10   NT x64  
WIN10.Company.Pri\SQL2017 WIN10        WIN10                       WIN10   NT x64  
SRV1\SQL2017              SRV1         SRV1                        SRV1    NT x64  
SRV1.Company.Pri\SQL2017  SRV1         SRV1                        SRV1    NT x64  

#>
```

I run this against the docker containers:
```
$cred = New-Object -TypeName PSCredential -ArgumentList "sqladmin", (ConvertTo-SecureString -String 'dbatools.IO' -AsPlainText -Force)
$DbaInstance = [DbaInstance[]]('localhost', 'localhost:1433', 'localhost:14333')
$DbaInstance | Format-Table
Connect-DbaInstance -SqlInstance $DbaInstance -SqlCredential $cred | Format-Table -Property Name, ComputerName, ComputerNamePhysicalNetBIOS, NetName, Platform

<# Output:

ComputerName InstanceName  Port NetworkProtocol IsLocalHost FullName        FullSmoName     SqlComputerName SqlInstanceName SqlFullName
------------ ------------  ---- --------------- ----------- --------        -----------     --------------- --------------- -----------
localhost    MSSQLSERVER   1433             Any        True localhost       localhost       [localhost]     [MSSQLSERVER]   [localhost]
localhost    MSSQLSERVER   1433             Any        True localhost:1433  localhost,1433  [localhost]     [MSSQLSERVER]   [localhost]
localhost    MSSQLSERVER  14333             Any        True localhost:14333 localhost,14333 [localhost]     [MSSQLSERVER]   [localhost]

Name            ComputerName ComputerNamePhysicalNetBIOS NetName    Platform
----            ------------ --------------------------- -------    --------
GALLOPER        localhost    dockersql1                  dockersql1 NT x64  
localhost,1433  localhost    dockersql1                  dockersql1 NT x64  
localhost,14333 localhost    dockersql2                  dockersql2 NT x64  

#>

```

Should that look like this?
* Platform for docker is "NT x64"
* The full qualified computer name is only in Name in the smo, not in ComputerName

In Get-DbaEndpoint there is this code:
https://github.com/sqlcollaborative/dbatools/blob/8db18ace16c3d23615f88673565b02eaf8a9856d/functions/Get-DbaEndpoint.ps1#L86-L95

Why is there a test `$instance.ComputerName -match '\.'` if ComputerName just never has a domain and so no dot in there?

Ok, Ok, a little bit unsorted, I apologize.

### PR DESCRIPTION
**Pull request by potatoqualitee**
## Type of Change

 - [x] Bug fix (non-breaking change, fixes #6814 )
 
I'd still like to know why this works. Can't figure it out yet. The issue is within Get-DbaEndPoint or New-DbaAgReplica which uses the endpoint.

Modifying Get-DbaEndPoint doesn't fix it, nor does modifying New-DbaAvailabilityGroup. This has gotta be an SMO thing. I'm so confused.

### CODE REVIEW

**Comment by wsmelton** [comment]
It works becuase you are modifying the `ComputerName` property that we set in `Connect-DbaInstance` to another value...*before* you call `Get-DbaEndpoint` which depends on that value being set to something.

**Comment by potatoqualitee** [comment]
right but what's weird is that if i remove it from add-dbaagreplica, and place it right under $server in get-dbaendpoint, it doesn't work! whattt?

**Comment by andreasjordan** [comment]
> right but what's weird is that if i remove it from add-dbaagreplica, and place it right under $server in get-dbaendpoint, it doesn't work! whattt?

Thats because $instance is used and not $server. That's a bug. (And I have not found it while working on Get-DbaEndpoint ...)

**Comment by potatoqualitee** [comment]
$instance is $server in this case. The dbainstance parameter type also supports SMO objects.

**Comment by wsmelton** [comment]
If I move it to `Get-DbaEndpoint` it works properly for me and is populating the ComputerName with the Container ID...what do you get?

image

**Comment by andreasjordan** [comment]
Yes, but $instance.ComputerName is used to build the Fqdn (but $server should have been used). Have you changed $instance or $server in Get-DbaEndpoint?

**Comment by potatoqualitee** [comment]
@wsmelton - it works properly for Get-DbaEndpoint, but somehow not for New-DbaAvailabiltyGroup. Sorry, should have been more clear. For New-DbaAvailabiltyGroup to work with Docker, it has to be in Add-DbaAgReplica.

**Comment by potatoqualitee** [comment]
Ohhh! I'll check that out, @andreasjordan !

**Comment by potatoqualitee** [comment]
I can see why I did that - basically, for this task, I trusted the user's input more than the way I built ComputerName. Let me think on this.

**Comment by potatoqualitee** [comment]
That was it! 🏆 Now to fix the FQDN problem you're experiencing with Connect-DbaInstance

**Comment by potatoqualitee** [comment]
Now it's in Get-DbaEndpoint, this feels much better.

File: functions/Get-DbaEndpoint.ps1 [hunk 1]
```diff
@@ -85,7 +91,7 @@ function Get-DbaEndpoint {
                 if ($end.Protocol.Tcp.ListenerPort) {
                     if ($end.Protocol.Tcp.ListenerIPAddress -ne [System.Net.IPAddress]'0.0.0.0') {
                         $dns = $end.Protocol.Tcp.ListenerIPAddress
-                    } elseif ($instance.ComputerName -match '\.') {
+                    } elseif ($instance.ComputerName -match '\.' -or $server.HostPlatform -eq "Linux" ) {
```
Comment (andreasjordan) [inline functions/Get-DbaEndpoint.ps1:94]: I think this is the real fix and the new lines 69-74 are not needed.

Comment (andreasjordan) [reply functions/Get-DbaEndpoint.ps1:94]: Ups, too early - that was completle wrong...

File: functions/Get-DbaEndpoint.ps1 [hunk 2]
```diff
@@ -66,6 +66,12 @@ function Get-DbaEndpoint {
                 Stop-Function -Message "Error occurred while establishing connection to $instance" -Category ConnectionError -ErrorRecord $_ -Target $instance -Continue
             }
 
+            # the next block is the best we can do for docker machines that don't usually resolve in DNS
+            # if this code block is placed after Get-DbaEndpoint, everything fails. Not sure why, yet.
+            if ($server.NetName -and $server.HostPlatform -eq "Linux" -and $instance.ComputerName -notmatch '\.') {
+                Add-Member -InputObject $instance -NotePropertyName ComputerName -NotePropertyValue $server.NetName -Force
```
Comment (andreasjordan) [inline functions/Get-DbaEndpoint.ps1:72]: What if you don't change $instance.ComputerName but change the if-statement down there with just another elseif ($server.NetName -and $server.HostPlatform -eq "Linux") { $dns = $server.NetName } ? That would be much clearer I think.

Comment (potatoqualitee) [reply functions/Get-DbaEndpoint.ps1:72]: It still get stuck. So weird. I'm guessing it has something to do with SMO and the way it adds replicas. 

image

**Pull request merged**